File: wrappers/javascript/shared/src/api/CredentialRequest.ts
import type { CredentialDefinition } from './CredentialDefinition'
import type { CredentialOffer } from './CredentialOffer'
import type { MasterSecret } from './MasterSecret'

import { IndyObject } from '../IndyObject'
import { indyCredx } from '../register'

import { CredentialRequestMetadata } from './CredentialRequestMetadata'

export type CreateCredentialRequestOptions = {
  proverDid: string
  credentialDefinition: CredentialDefinition
  masterSecret: MasterSecret
  masterSecretId: string
  credentialOffer: CredentialOffer
}

export class CredentialRequest extends IndyObject {
  public static create(options: CreateCredentialRequestOptions) {
    const { credentialRequest, credentialRequestMeta } = indyCredx.createCredentialRequest({
      proverDid: options.proverDid,
      credentialDefinition: options.credentialDefinition.handle,
      masterSecret: options.masterSecret.handle,
      masterSecretId: options.masterSecretId,
      credentialOffer: options.credentialOffer.handle,
    })

    return {
      credentialRequest: new CredentialRequest(credentialRequest.handle),
      credentialRequestMetadata: new CredentialRequestMetadata(credentialRequestMeta.handle),
    }
  }

  public static load(json: string) {
    return new CredentialRequest(indyCredx.credentialRequestFromJson({ json }).handle)
  }
}

File: wrappers/dotnet/lib/Requests/CredentialRequest.cs
using AnonCredsNet.Exceptions;
using AnonCredsNet.Interop;
using AnonCredsNet.Models;

namespace AnonCredsNet.Requests;

public class CredentialRequest : AnonCredsObject
{
    private CredentialRequest(long handle)
        : base(handle) { }

    public static (CredentialRequest Request, CredentialRequestMetadata Metadata) Create(
        CredentialDefinition credDef,
        string linkSecret,
        string linkSecretId,
        CredentialOffer credOffer,
        string? entropy = null,
        string? proverDid = null
    )
    {
        if (
            credDef == null
            || string.IsNullOrEmpty(linkSecret)
            || string.IsNullOrEmpty(linkSecretId)
            || credOffer == null
        )
            throw new ArgumentNullException("Input parameters cannot be null or empty");
        var code = NativeMethods.anoncreds_create_credential_request(
            entropy,
            proverDid,
            credDef.Handle,
            linkSecret,
            linkSecretId,
            credOffer.Handle,
            out var req,
            out var meta
        );
        if (code != ErrorCode.Success)
            throw new AnonCredsException(code, AnonCreds.GetCurrentError());
        return (new CredentialRequest(req), new CredentialRequestMetadata(meta));
    }

    internal static CredentialRequest FromJson(string json) => FromJson<CredentialRequest>(json);
}
